src/lib.rs

```diff
@@ -0,0 +1,507 @@
+    fn record_hierarchical(
+        &self,
+        dispatch: &Dispatch,
+        id: &span::Id,
+        f: WithContextCallback<'_>,
+        field: &dyn Fn() -> String,
+        panic: bool,
+    ) -> Option<(span::Id, Meta, field::Field)> {
+        (self.0)(dispatch, id, f, field, panic)
+    }
+}
+
+/// A [`Layer`] that helps [`field`]s find their corresponding [`Span`] in the
+/// chain of [`Span`]s.
+#[derive(Debug)]
+pub struct HierarchicalRecord<S> {
+    /// Type-erased [`Layer`]'s context.
+    ctx: WithContext,
+
+    /// [`Subscriber`] marker.
+    _marker: PhantomData<fn(S)>,
+}
+
+impl<S> Default for HierarchicalRecord<S>
+where
+    S: for<'span> LookupSpan<'span> + Subscriber,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<S> HierarchicalRecord<S>
+where
+    S: for<'span> LookupSpan<'span> + Subscriber,
+{
+    /// Creates new [`HierarchicalRecord`].
+    #[must_use]
+    pub fn new() -> Self {
+        Self { ctx: WithContext(Self::with_context), _marker: PhantomData }
+    }
+
+    /// Function saved and called by [`WithContext`] at the callsite.
+    #[allow(clippy::unwrap_in_result)]
+    fn with_context(
+        dispatch: &Dispatch,
+        id: &span::Id,
+        f: WithContextCallback<'_>,
+        field: &dyn Fn() -> String,
+        panic: bool,
+    ) -> Option<(span::Id, Meta, field::Field)> {
+        let subscriber =
+            dispatch.downcast_ref::<S>().expect("subscriber not found");
+        let span = subscriber.span(id).expect("unknown span");
+
+        #[allow(clippy::option_if_let_else)]
+        if let Some(parent) = span.parent() {
+            f(parent.id(), parent.metadata())
+        } else {
+            // `Span` wants to record a field but has no parents. This means
+            // that we walked entire chain of `Span`s to the root, yet this
+            // field did not find it's corresponding `Span`. We know that
+            // because otherwise the iteration in `record_parent` would end, and
+            // this function would not be called again anymore, yet it is.
+            // Nothing to do but report the error.
+
+            let meta = span.metadata();
+            let field = field();
+            // We log and then panic to avoid situation, when we get double
+            // panic without any info.
+            log::error!(
+                "Span(id={id:?}, meta={meta:?}) doesn't have `{field}` field"
+            );
+            panic.then(|| {
+                panic!(
+                    "Span(id={id:?}, meta={meta:?}) doesn't have `{field}` \
+                     field"
+                )
+            })
+        }
+    }
+}
+
+impl<S> Layer<S> for HierarchicalRecord<S>
+where
+    S: for<'span> LookupSpan<'span> + Subscriber,
+{
+    // SAFETY: This is safe because the `WithContext` function pointer is valid
+    //         for the lifetime of `&self`.
+    #[allow(unsafe_code)]
+    unsafe fn downcast_raw(&self, id: TypeId) -> Option<*const ()> {
```

Comment:
@zohnannor with a minor priority, think other way around this `unsafe` impl, so we don't introduce it in our crate.